### PR TITLE
fix(minifier): break directive prologue

### PR DESCRIPTION
**Pull request by mysteryven**
## What's happened?

`EmptyStatement` is not safety to drop when it is followed by a `StringLiteral`:

```js
function foo() {
	;'use strict'
}
```

Terser just stay it in place:

```js
function t(){;"use strict"}
```

## Terser

1. https://github.com/terser/terser/blob/28bc97af0d7ef6606c0ed4a7a2403362b6c2a0e8/lib/output.js#L1215
2. https://github.com/terser/terser/blob/28bc97af0d7ef6606c0ed4a7a2403362b6c2a0e8/lib/output.js#L845

## One more thing

It is not always keep `;`.

```js
function foo() {
  ;'use strict';
  console.log('hi')
}
```

It will be compressed to :

```js
function o(){"use strict",console.log("hi")};
```

Maybe I can continue to implement this condition when has test cases.

### CODE REVIEW

File: crates/oxc_minifier/src/printer/gen.rs [hunk 1]
```diff
@@ -502,7 +502,13 @@ impl<'a> Gen for FunctionBody<'a> {
         for directive in &self.directives {
             directive.gen(p);
         }
-        p.needs_semicolon = false;
+        p.needs_semicolon =
+            if let Some(Statement::ExpressionStatement(expr_stmt)) = self.statements.get(0)
+            && let Expression::StringLiteral(_) = &expr_stmt.expression {
```
Comment (Boshen) [inline crates/oxc_minifier/src/printer/gen.rs:507]: `self.statements.get(0).filter(|expr| matches!(expr, ...)`

**Review by Boshen — CHANGES_REQUESTED** [review]

File: crates/oxc_minifier/src/printer/gen.rs (hunk 1 above)
Comment (Boshen) [reply crates/oxc_minifier/src/printer/gen.rs:507]: Oh, this is not going to work with box_patterns

then try `&& matches!(expr_stmt.expression, ...)`

File: crates/oxc_minifier/src/printer/gen.rs [hunk 2]
```diff
@@ -502,7 +502,13 @@ impl<'a> Gen for FunctionBody<'a> {
...
+                true
+            } else {
+                false
+            };
         for stmt in &self.statements {
```
Comment (Boshen) [inline crates/oxc_minifier/src/printer/gen.rs:512]: you get add an index here, self.statements.iter().enumerate(), then check i == 0 and stmt == string

Comment (mysteryven) [reply crates/oxc_minifier/src/printer/gen.rs:512]: I wonder if only checking once in the outer of for loop is a more effective way.

Comment (Boshen) [reply crates/oxc_minifier/src/printer/gen.rs:512]: > I wonder if only checking once in the outer of for loop is a more effective way.

Good point, both are fine by me.

**Review by Boshen — APPROVED** [review]

**Pull request merged**